From baa19e40d997cdd75ceefd99bb5c4677fd4f260f Mon Sep 17 00:00:00 2001
From: Steve Dickson <steved@redhat.com>
Date: Wed, 16 Mar 2011 13:58:15 -0400
Subject: [PATCH 1/1] Make superuser the owner of registers from the kernel.

Registers from the kernel do not come in on AF_LOCAL
sockets so the exact owners of the service can not be determined.
But given the fact the loopback address is used and a privilege
port used, it can be assumed the owner of the service should
be the superuser.

Signed-off-by: Steve Dickson <steved@redhat.com>
---
 src/rpcb_svc_com.c |   11 +++++++----
 1 files changed, 7 insertions(+), 4 deletions(-)

diff --git a/src/rpcb_svc_com.c b/src/rpcb_svc_com.c
index 291421f..5bb9a44 100644
--- a/src/rpcb_svc_com.c
+++ b/src/rpcb_svc_com.c
@@ -1372,10 +1372,13 @@ static char *
 getowner(SVCXPRT *transp, char *owner, size_t ownersize)
 {
 	uid_t uid;
- 
-	if (__rpc_get_local_uid(transp, &uid) < 0)
-		snprintf(owner, ownersize, "unknown");
-	else if (uid == 0)
+
+	if (__rpc_get_local_uid(transp, &uid) < 0) {
+		if (is_localroot(svc_getrpccaller(transp)))
+			snprintf(owner, ownersize, "superuser");
+		else 
+			snprintf(owner, ownersize, "unknown");
+	} else if (uid == 0)
 		snprintf(owner, ownersize, "superuser");
 	else
 		snprintf(owner, ownersize, "%d", uid);  
-- 
1.7.7.6

